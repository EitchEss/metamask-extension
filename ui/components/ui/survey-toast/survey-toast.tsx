import React, { useEffect, useState, useContext } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import fetchWithCache from '../../../../shared/lib/fetch-with-cache';
import { DAY } from '../../../../shared/constants/time';
import { MetaMetricsContext } from '../../../contexts/metametrics';
import {
  MetaMetricsEventCategory,
  MetaMetricsEventName,
} from '../../../../shared/constants/metametrics';
import {
  getSelectedInternalAccount,
  getLastViewedUserSurvey,
  getUseExternalServices,
  getParticipateInMetaMetrics,
} from '../../../selectors';
import { setLastViewedUserSurvey } from '../../../store/actions';
import { Toast } from '../../multichain';

type Survey = {
  url: string;
  description: string;
  cta: string;
  surveyId: number;
};

export function SurveyToast() {
  const [survey, setSurvey] = useState<Survey | null>(null);
  const dispatch = useDispatch();
  const trackEvent = useContext(MetaMetricsContext);
  const lastViewedUserSurvey = useSelector(getLastViewedUserSurvey);
  const participateInMetaMetrics = useSelector(getParticipateInMetaMetrics);
  const basicFunctionality = useSelector(getUseExternalServices);
  const internalAccount = useSelector(getSelectedInternalAccount);

  useEffect(() => {
    if (!basicFunctionality) {
      return;
    }

    const surveyId = 1;
    const surveyUrl = `https://accounts.dev-api.cx.metamask.io/v1/users/${internalAccount.address}/surveys?surveyId=${surveyId}`;

    const fetchSurvey = async () => {
      try {
        const response = await fetchWithCache({
          url: surveyUrl,
          fetchOptions: {
            method: 'GET',
            headers: {
              'x-metamask-clientproduct': 'metamask-extension',
            },
          },
          functionName: 'fetchSurveys',
          cacheOptions: { cacheRefreshTime: DAY * 7 },
        });

        const _survey: Survey = response?.surveys?.[0];

        if (
          response.surveys.length === 0 ||
          !_survey ||
          _survey.surveyId <= lastViewedUserSurvey
        ) {
          return;
        }

        setSurvey(_survey);
      } catch (error) {
        console.error('Failed to fetch survey:', error);
      }
    };

    fetchSurvey();
  }, [
    internalAccount.address,
    lastViewedUserSurvey,
    basicFunctionality,
    dispatch,
  ]);

  function handleActionClick() {
    if (!survey) {
      return;
    }
    window.open(survey.url, '_blank');
    dispatch(setLastViewedUserSurvey(survey.surveyId));
    trackAction('accept');
  }

  function handleClose() {
    if (!survey) {
      return;
    }
    dispatch(setLastViewedUserSurvey(survey.surveyId));
    trackAction('deny');
  }

  function trackAction(response: 'accept' | 'deny') {
    if (!participateInMetaMetrics || !survey) {
      return;
    }

    trackEvent({
      event: MetaMetricsEventName.SurveyToast,
      category: MetaMetricsEventCategory.Feedback,
      properties: {
        response,
        survey: survey.surveyId,
      },
    });
  }

  if (!survey) {
    return null;
  }

  return (
    <Toast
      dataTestId="survey-toast"
      key="survey-toast"
      text={survey.description}
      actionText={survey.cta}
      onActionClick={handleActionClick}
      onClose={handleClose}
      startAdornment={undefined}
    />
  );
}
