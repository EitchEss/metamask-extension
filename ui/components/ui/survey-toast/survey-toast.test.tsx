import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchWithCache from '../../../../shared/lib/fetch-with-cache';
import { MetaMetricsContext } from '../../../contexts/metametrics';
import { SurveyToast } from './survey-toast';

jest.mock('../../../../shared/lib/fetch-with-cache', () => ({
  __esModule: true,
  default: jest.fn(),
}));

window.open = jest.fn();

const mockFetchWithCache = fetchWithCache as jest.Mock;
const mockStore = configureStore([thunk]);

const initialState = {
  metametrics: {
    participateInMetaMetrics: true,
  },
  user: {
    basicFunctionality: true,
  },
  metamask: {
    lastViewedUserSurvey: 0,
    useExternalServices: true,
    internalAccounts: {
      selectedAccount: '0x123',
      accounts: { '0x123': { address: '0x123' } },
    },
  },
};

const store = mockStore(initialState);
const mockTrackEvent = jest.fn();

const renderComponent = () =>
  render(
    <Provider store={store}>
      <MetaMetricsContext.Provider value={mockTrackEvent}>
        <SurveyToast />
      </MetaMetricsContext.Provider>
    </Provider>,
  );

describe('SurveyToast', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  afterEach(() => {
    jest.resetAllMocks();
    jest.restoreAllMocks();
  });

  it('should match snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('renders nothing if no survey is available', () => {
    mockFetchWithCache.mockResolvedValue({ surveys: [] });
    renderComponent();
    expect(screen.queryByTestId('survey-toast')).toBeNull();
  });

  it('renders the survey toast if a survey is available', async () => {
    const survey = {
      url: 'https://example.com',
      description: 'Test Survey',
      cta: 'Take Survey',
      surveyId: 1,
    };
    mockFetchWithCache.mockResolvedValue({ surveys: [survey] });

    renderComponent();

    await waitFor(() => {
      expect(screen.getByTestId('survey-toast')).toBeInTheDocument();
      expect(screen.getByText('Test Survey')).toBeInTheDocument();
      expect(screen.getByText('Take Survey')).toBeInTheDocument();
    });
  });

  it('handles action click correctly', async () => {
    const survey = {
      url: 'https://example.com',
      description: 'Test Survey',
      cta: 'Take Survey',
      surveyId: 1,
    };
    mockFetchWithCache.mockResolvedValue({ surveys: [survey] });

    renderComponent();

    await waitFor(() => {
      expect(screen.getByTestId('survey-toast')).toBeInTheDocument();
    });

    fireEvent.click(screen.getByText('Take Survey'));

    expect(window.open).toHaveBeenCalledWith('https://example.com', '_blank');
  });
});
