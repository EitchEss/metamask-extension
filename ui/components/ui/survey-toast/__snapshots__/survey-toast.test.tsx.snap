// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SurveyToast should match snapshot 1`] = `
<div>
  <div
    class="mm-box mm-banner-base mm-box--padding-3 mm-box--display-flex mm-box--gap-2 mm-box--background-color-background-default"
    data-theme="light"
  >
    <div
      class="mm-box mm-box--min-width-0"
    >
      <div
        class="mm-box mm-box--display-flex mm-box--gap-4"
        data-testid="survey-toast"
      >
        <div
          class="mm-box"
        >
          <p
            class="mm-box mm-text toast-text mm-text--body-md mm-box--color-text-default"
          />
        </div>
      </div>
    </div>
    <button
      aria-label="Close"
      class="mm-box mm-button-icon mm-button-icon--size-sm mm-banner-base__close-button mm-box--margin-left-auto mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-icon-default mm-box--background-color-transparent mm-box--rounded-lg"
    >
      <span
        class="mm-box mm-icon mm-icon--size-sm mm-box--display-inline-block mm-box--color-inherit"
        style="mask-image: url('./images/icons/close.svg');"
      />
    </button>
  </div>
</div>
`;
